### PR TITLE
uninterleave

### PR DESCRIPTION
**Pull request by BlackCapCoder**
I am agnostic on the names!

### CODE REVIEW

**Review by Bodigrim — COMMENTED** [review]
Is there any prior art or applications for these functions?

**Comment by BlackCapCoder** [comment]
```haskell
data Infinite a = a :< Infinite a
  deriving Foldable

infixr 5 :<

-- | Interleave two infinite lists.
interleave1 :: Infinite a -> Infinite a -> Infinite a
interleave1 (l :< ls) ~(r :< rs) = l :< r :< interleave1 ls rs

interleave2 :: Infinite a -> Infinite a -> Infinite a
interleave2 (x :< xs) ys = x :< interleave2 ys xs

nats = go 1 where
  go n = n :< go (succ n)

test s = Prelude.toList s !! 1_000_000_0

λ :set +s
λ test $ interleave1 nats nats
5000001
(4.90 secs, 3,169,926,688 bytes)
λ test $ interleave2 nats nats
5000001
(5.49 secs, 3,609,926,744 bytes)
```

**Comment by BlackCapCoder** [comment]
> Is there any prior art or applications for these functions?

You can use `uninterleave` like you would use `Supply` from streams. It also has the nice property that stitching streams together with these primitives is fully reversible, and this property is unique to infinite streams!

**Comment by Bodigrim** [comment]
I can find [limited prior art for `uninterleave`](https://hackage-search.serokell.io/?q=uninterleave), but not for `interswap`. I appreciate your contribution, but let's leave `interswap` aside. We cannot really provide all possible helpers and combinations. 

With regards to `uninterleave`, could you please make it lazier? E. g., `head (fst (uninterleave ('a' :< undefined)))` should succeed. 

Performance measurements in `ghci` are typically unreliable and unrepresentative for compiled code. If you want to make a case about performance, may I suggest you to write a benchmark suite using `tasty-bench`?

**Comment by BlackCapCoder** [comment]
The old `interleave` implementation seems to be faster in compiled code!

**Comment by BlackCapCoder** [comment]
[I wrote some benchmarks for `uninterleave`](https://gist.github.com/BlackCapCoder/9532582173b487ab0c073bd906042d0e). They seem to support my statements about performance from the comments, and we can gain about 30% by implementing `uninterleaveR` in terms of `uninterleaveL`.

However, I am not happy about "warmup dummy" in my benchmarks! Perhaps you could offer some insight?

**Comment by Bodigrim** [comment]
I like your `uninterleave3` the most: it is better not to clutter the top level name space. You can make it even sweeter, expressing `uninterleaveL` via `uninterleaveR` and vice versa, avoiding irrefutable pattern matching.

**Comment by Bodigrim** [comment]
> It is half the speed of uninterleave1 though. That's a very steep price for not having top level definitions, that we don't export anyways

I'm sorry, I don't follow. The benchmarks you quoted above (https://gist.github.com/BlackCapCoder/9532582173b487ab0c073bd906042d0e) show that `uninterleave3` is the fastest one, slightly faster than `uninterleave1`. How come that it's now half the speed? 

> edit: I tried expressing uninterleaveL and uninterleaveR in terms of each other, and it's worse.

Can you give me a reproducible benchmark for this? Please make it a proper benchmark suite which I can run with `cabal bench`.

**Comment by Bodigrim** [comment]
I've added an empty benchmark suite, feel free to populate it.

**Comment by BlackCapCoder** [comment]
I added some benchmarks; I think I've just been looking at the wrong number. I've been looking at the total time!

`uninterleave7` is with local definitions and L/R in terms of each other
```
All
  uninterleave1: OK (3.59s)
    125  ms ± 9.3 ms, 512 MB allocated,  25 KB copied, 1.4 GB peak memory
  uninterleave2: OK (1.54s)
    120  ms ± 4.5 ms, 512 MB allocated,  24 KB copied, 1.4 GB peak memory
  uninterleave3: OK (4.05s)
    405  ms ±  28 ms, 2.2 GB allocated, 116 KB copied, 1.5 GB peak memory
  uninterleave4: OK (1.65s)
    417  ms ±  10 ms, 2.2 GB allocated, 112 KB copied, 1.5 GB peak memory
  uninterleave5: OK (2.12s)
    440  ms ±  12 ms, 2.5 GB allocated, 114 KB copied, 1.5 GB peak memory
  uninterleave6: OK (1.71s)
    125  ms ± 9.3 ms, 512 MB allocated,  20 KB copied, 1.5 GB peak memory
  uninterleave7: OK (0.70s)
    116  ms ± 5.4 ms, 512 MB allocated,  19 KB copied, 1.5 GB peak memory
```

File: bench/Uninterleave.hs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE ViewPatterns, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+module Uninterleave where
+
+import Test.Tasty.Bench
+import Data.List.Infinite (Infinite (..))
+
+
+benchmarks :: [Benchmark]
+benchmarks =
```
Comment (Bodigrim) [inline bench/Uninterleave.hs:10]:
```suggestion
benchmarks :: Benchmark
benchmarks = bgroup "uninterleave"
```

File: bench/Bench.hs [hunk 2]
```diff
@@ -4,8 +4,12 @@ module Main where
...
+import qualified Uninterleave
+
 
 main :: IO ()
-main = defaultMain
+main = defaultMain $
   [
+  ] <> concat
+  [ Uninterleave.benchmarks
   ]
```
Comment (Bodigrim) [inline bench/Bench.hs:15]:
```suggestion
main = defaultMain
  [ Uninterleave.benchmarks
  ]
```
As witnessed by CI, old versions of GHC are unhappy with `(<>)`.

**Review by Bodigrim — COMMENTED** [review]
Overall looks good, but let's make CI happy.

**Comment by Bodigrim** [comment]
Oops, this was unintentional. Feel free to rebase and reopen.

**Pull request closed without merging**